### DISCUSSION TITLE
Ferrum on Ruby 3.1 with "Browser is dead or given window is closed" error and jemalloc

### DISCUSSION THREAD

**Discussion by jweir** [rubycdp/ferrum#248] [General]
### Update

After further review the problem appears to be in the `google-chrome` application when `jemalloc` is enabled.

My previous configuration was running Ruby 2.7 with jemalloc compiled into Ruby.  When switching to Ruby 3.1 jemalloc was enabled via the LD_PRELOAD env.  This caused a global change that impacted more than just Ruby.

### Original

I am not sure where to put this, so I will discuss it here in the even another person comes across this issue.

After upgrading to Ruby 3.1 from 2.7 our Ferrum processes stopped working. 

```
Ferrum::Browser.new.go_to("https://google.com")
```
Would error with
```
/home/deploy/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/ferrum-0.11/lib/ferrum/browser/client.rb:45:in `command': Browser is dead or given window is closed (Ferrum::DeadBrowserError)
	from /home/deploy/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/ferrum-0.11/lib/ferrum/page.rb:160:in `command'
	from /home/deploy/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/ferrum-0.11/lib/ferrum/page.rb:72:in `go_to'
	from /home/deploy/.rbenv/versions/3.1.0/lib/ruby/3.1.0/forwardable.rb:238:in `go_to'
	from (irb):3:in `<main>'
	from /home/deploy/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
	from /home/deploy/.rbenv/versions/3.1.0/bin/irb:25:in `load'
	from /home/deploy/.rbenv/versions/3.1.0/bin/irb:25:in `<main>'
```

At the same time in `syslog` the below error appeared
```
Feb 19 00:04:39 app kernel: [18384.952998] NetworkService[2765]: segfault at 4210 ip 00007fb216a1d7f4 sp 00007fb2011d4610 error 4 in libjemalloc.so.2[7fb2169a5000+8b000]
```

Our Ruby instances were running with `jemalloc-5.2.1` via the env var `LD_PRELOAD="/usr/local/lib/libjemalloc.so"` on Ubuntu 18.04.

Running without `jemalloc` has allowed our ferrum sessions to run again.

**Comment by route** [comment]
Unfortunately I think it's better to open an issue in ruby itself, do you have core dump?

**Reply by jweir** [reply]
No core dump, not even sure how to get one since it appears to be some forked, or threaded process that is segfaulting – the running process just has the error.

I have just tested this in 3.1.1 and it still errors.

I am going to open a ticket in the Ruby tracker.

**Reply by jweir** [reply]
Nevermind.  I  am confident this is a google-chome issue with jemalloc.  I am unable to run it with jemalloc enabled, but it runs without.

```
vagrant@app:~$ export LD_PRELOAD=/usr/local/lib/libjemalloc.so
vagrant@app:~$ google-chrome
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
[0224/195912.378655:ERROR:file_io_posix.cc(144)] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq: No such file or directory (2)
[0224/195912.378702:ERROR:file_io_posix.cc(144)] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: No such file or directory (2)
Segmentation fault (core dumped)
```
